File: anime.mjs
import fs from 'node:fs';

const MAX_BYTE = 255;
const framebufferPath = '/dev/fb0';

const Screen = () => {
    const ans = {};
    ans.width = 2560;
    ans.height = 1440;
    ans.bytesPerPixel = 4;
    ans.channels = 4;
    ans.size = ans.width * ans.height * ans.bytesPerPixel;
    ans.buffer = Buffer.allocUnsafe(ans.size);
    ans.fb = fs.openSync(framebufferPath, 'r+');
    ans.map = (lambda, time) => {
        const channels = ans.bytesPerPixel;
        const w = ans.width;
        const h = ans.height;
        for (let k = 0; k < ans.size; k += channels) {
            const i = k / (channels * w);
            const j = (k / channels) % w;
            const x = j;
            const y = h - 1 - i;
            const color = lambda(x, y, time);
            if (!color) continue;
            ans.buffer[k] = MAX_BYTE * color[2];
            ans.buffer[k + 1] = MAX_BYTE * color[1];
            ans.buffer[k + 2] = MAX_BYTE * color[0];
            ans.buffer[k + 3] = MAX_BYTE;
        }
        return ans.paint();
    }
    ans.paint = () => {
        fs.writeSync(ans.fb, ans.buffer, 0, ans.size, 0);
        return ans;
    }
    return ans;
}

const screen = Screen();
const render = (x, y, t) => {
    const px = (x * t) / screen.width;
    const py = (y * t) / screen.height;
    return [px % 1, py % 1, 0];
}

const play = ({ oldTime, time }) => {
    const newTime = new Date().getTime();
    const dt = (newTime - oldTime) * 1e-3;
    screen.map(render, time);
    setTimeout(() => play({ oldTime: newTime, time: time + dt }));
}
play({ oldTime: new Date().getTime(), time: 0 });

File: anime.c
#include <fcntl.h>
#include <linux/fb.h>
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/ioctl.h>
#include <sys/mman.h>
#include <unistd.h>

#define BYTE 8
#define MAX_BYTE 255

typedef struct screen_struct {
  uint32_t size;
  uint32_t width;
  uint32_t height;
  uint8_t *canvas;
  uint8_t channels;
  int frame_buffer_fd;
} Screen;

Screen *new_screen() {
  struct fb_var_screeninfo vinfo;

  int fb_fd = open("/dev/fb0", O_RDWR);
  if (fb_fd == -1) {
    perror("Error: cannot open framebuffer device");
    exit(1);
  }

  if (ioctl(fb_fd, FBIOGET_VSCREENINFO, &vinfo) == -1) {
    perror("Error reading variable information");
    close(fb_fd);
    exit(2);
  }

  int w = vinfo.xres;
  int h = vinfo.yres;
  int color_channels = vinfo.bits_per_pixel / BYTE;
  long screen_size = w * h * color_channels;

  printf("Framebuffer size: width=%d, height=%d, channels=%d\n", w, h,
         color_channels);

  // We avoid mmap for double buffering simulation
  uint8_t *canvas = (uint8_t *)malloc(sizeof(uint8_t) * screen_size);
  if (!canvas) {
    perror("Failed to allocate canvas buffer");
    close(fb_fd);
    exit(3);
  }

  Screen *ans = malloc(sizeof(Screen));
  ans->width = w;
  ans->height = h;
  ans->canvas = canvas;
  ans->size = screen_size;
  ans->channels = color_channels;
  ans->frame_buffer_fd = fb_fd;

  return ans;
}

void close_screen(Screen *screen) {
  free(screen->canvas);
  close(screen->frame_buffer_fd);
  free(screen);
}

Screen *map(Screen *screen,
            uint8_t *(*lambda)(Screen *s, uint32_t x, uint32_t y, void *c),
            void *context) {
  uint32_t screen_size = screen->size;
  uint32_t w = screen->width;
  uint32_t h = screen->height;
  uint8_t channels = screen->channels;

  for (uint32_t k = 0; k < screen_size; k += channels) {
    uint32_t i = k / (channels * w);
    uint32_t j = (k / channels) % w;
    uint32_t x = j;
    uint32_t y = h - 1 - i;
    uint8_t *color = lambda(screen, x, y, context);
    if (color == NULL)
      continue;
    screen->canvas[k] = color[2];     // Blue
    screen->canvas[k + 1] = color[1]; // Green
    screen->canvas[k + 2] = color[0]; // Red
    if (channels == 4) {
      screen->canvas[k + 3] = MAX_BYTE; // Alpha (if present)
    }
    free(color);
  }
  // 	Reset file pointer to start of framebuffer
  lseek(screen->frame_buffer_fd, 0, SEEK_SET);
  // Push whole canvas to framebuffer in one call
  write(screen->frame_buffer_fd, screen->canvas, screen->size);

  return screen;
}



typedef struct time_struct {
  double time;
} Time;

uint8_t *anime(Screen *screen, uint32_t x, uint32_t y, void *context) {
  uint8_t *ans = malloc(sizeof(uint8_t) * 3);
  double time = ((Time *)context)->time;

  double px = (double)x * time / screen->width;
  double py = (double)y * time / screen->height;

  ans[0] = ((uint8_t)(MAX_BYTE * px)) % MAX_BYTE;
  ans[1] = ((uint8_t)(MAX_BYTE * py)) % MAX_BYTE;
  ans[2] = (uint8_t)0;

  return ans;
}

int main() {
  Screen *screen = new_screen();
  double t = 0;

  while (t < 10.0) {
    Time time = {.time = t};
    map(screen, anime, &time);
    t += 0.01;
  }

  close_screen(screen);
  return 0;
}
